File: basic/src/main/java/com/yuanshenbin/basic/config/BasicOptions.java
package com.yuanshenbin.basic.config;

import android.content.Context;

import com.tencent.mmkv.MMKV;
import com.yuanshenbin.basic.call.ICatchConfig;
import com.yuanshenbin.basic.develop.DevelopConfig;
import com.yuanshenbin.basic.develop.DevelopMode;
import com.yuanshenbin.basic.develop.ui.ErrorDescActivity;
import com.yuanshenbin.basic.imgloader.IImageLoaderProxy;
import com.yuanshenbin.basic.imgloader.ImageLoader;
import com.yuanshenbin.basic.log.ILog;
import com.yuanshenbin.basic.log.LogAbstract;
import com.yuanshenbin.basic.log.LogImpl;
import com.yuanshenbin.basic.util.SPProxy;
import com.yuanshenbin.basic.util.SPUtils;

import cat.ereza.customactivityoncrash.config.CaocConfig;

/**
 * author : yuanshenbin
 * time   : 2021-01-25
 * desc   : 一些基础配置
 */
public class BasicOptions {
    private int pageSize = 20;//分页默认值

    private int compressSize=100;
    private int compressCount=10;
    private static BasicOptions instance;
    private Context mContext;
    private IImageLoaderProxy mIImageLoaderProxy;
    private boolean debug;

    private DevelopConfig developConfig;
    private ICatchConfig catchConfig;

    private LogAbstract logAbstract = new LogImpl();


    private TipsAbstract mTipsAbstract = new TipsAbstract() {
    };
    private static MMKV mMMKV;
    private static SPProxy mSPProxy;


    public static BasicOptions getInstance() {

        if (instance == null)
            synchronized (BasicOptions.class) {
                instance = new BasicOptions();
            }
        return instance;
    }

    public BasicOptions init(Context context, boolean debug) {
        this.mContext = context;
        this.debug = debug;
        return this;
    }

    public BasicOptions spProxy(SPProxy spProxy) {
        this.mSPProxy = spProxy;
        return this;
    }

    public BasicOptions compressSize(int compressSize) {
        this.compressSize = compressSize;
        return this;
    }
    public BasicOptions compressCount(int compressCount) {
        this.compressCount = compressCount;
        return this;
    }


    public BasicOptions tipsDialogConfig(TipsAbstract tipsAbstract) {
        this.mTipsAbstract = tipsAbstract;
        return this;
    }

    public BasicOptions developConfig(DevelopConfig developConfig) {
        this.developConfig = developConfig;
        return this;
    }

    public BasicOptions logAbstract(LogAbstract logAbstract) {
        this.logAbstract = logAbstract;
        return this;
    }

    public BasicOptions catchConfig(ICatchConfig catchConfig) {
        this.catchConfig = catchConfig;
        return this;
    }


    public BasicOptions imageLoaderProxy(IImageLoaderProxy iImageLoaderProxy) {
        this.mIImageLoaderProxy = iImageLoaderProxy;
        return this;
    }

    public void build() {
        ILog.initialize(logAbstract);


        if (catchConfig != null) {


            catchConfig.config(new CaocConfig.Builder()
                    .errorActivity(ErrorDescActivity.class).trackActivities(true));
        } else {

            CaocConfig.Builder.create().errorActivity(ErrorDescActivity.class)
                    .trackActivities(true).apply();
        }

        if (mIImageLoaderProxy != null) {
            ImageLoader.getInstance()
                    .setImageLoaderProxy(mIImageLoaderProxy);
        }
        if (developConfig != null) {
            DevelopMode
                    .getInstance()
                    .InitializationConfig(developConfig);
        }
        if (mSPProxy != null) {
            SPUtils.initialize(mSPProxy);
        } else {
            mMMKV = MMKV.mmkvWithID(getContext().getPackageName());
            SPUtils.initialize(new SPProxy() {
                @Override
                public MMKV getSP(String key) {
                    return mMMKV;
                }
            });
        }
    }

    public TipsAbstract getTipsAbstract() {
        return mTipsAbstract;
    }

    public Context getContext() {
        return mContext;
    }

    public int getPageSize() {
        return pageSize;
    }

    public void pageSize(int pageSize) {
        this.pageSize = pageSize;
    }

    public int getCompressSize() {
        return compressSize;
    }

    public int getCompressCount() {
        return compressCount;
    }

    public boolean isDebug() {
        return debug;
    }
}

File: basic-kotlin/src/main/java/com/yuanshenbin/basic/config/BasicOptions.kt
package com.yuanshenbin.basic.config

import android.content.Context
import com.tencent.mmkv.MMKV
import com.yuanshenbin.basic.develop.DevelopConfig
import com.yuanshenbin.basic.develop.DevelopMode
import com.yuanshenbin.basic.imgloader.IImageLoaderProxy
import com.yuanshenbin.basic.imgloader.ImageLoader
import com.yuanshenbin.basic.log.ILog
import com.yuanshenbin.basic.log.LogAbstract
import com.yuanshenbin.basic.log.LogImpl
import com.yuanshenbin.basic.util.SPProxy
import com.yuanshenbin.basic.util.SPUtils

/**
 * author : yuanshenbin
 * time   : 2021-01-25
 * desc   : 一些基础配置
 */
class BasicOptions {
    var pageSize = 20 //分页默认值
        private set
    var context: Context? = null
        private set
    var compressSize = 100
        //默认压缩大小
        private set
    var compressCount = 10
        //默认压缩次数
        private set
    private var mMMKV: MMKV? = null
    private val mIImageLoaderProxy: IImageLoaderProxy? = null
    var tipsAbstract: TipsAbstract = object : TipsAbstract() {}
        private set
    var spProxy: SPProxy? = null
        private set
    private var debug: Boolean? = false
    var developConfig: DevelopConfig? = null
    var logAbstract: LogAbstract = object : LogImpl() {}

    fun init(context: Context?, debug: Boolean): BasicOptions {
        this.context = context
        this.debug = debug;
        return this
    }

    fun isDebug(): Boolean? {
        return debug
    }

    fun logAbstract(abstract: LogAbstract?): BasicOptions {
        this.logAbstract = abstract!!
        return this
    }

    fun spProxy(spProxy: SPProxy?): BasicOptions {
        this.spProxy = spProxy
        return this
    }

    fun tipsDialogConfig(tipsAbstract: TipsAbstract): BasicOptions {
        this.tipsAbstract = tipsAbstract
        return this
    }

    fun developConfig(developConfig: DevelopConfig): BasicOptions {
        this.developConfig = developConfig
        return this
    }

    fun imageLoaderProxy(iImageLoaderProxy: IImageLoaderProxy?): BasicOptions {
        ImageLoader.instance
                .setImageLoaderProxy(iImageLoaderProxy)
        return this
    }

    fun pageSize(pageSize: Int): BasicOptions {
        this.pageSize = pageSize
        return this
    }

    fun compressSize(compressSize: Int): BasicOptions {
        this.compressSize = compressSize
        return this
    }

    fun compressCount(compressCount: Int): BasicOptions {
        this.compressCount = compressCount
        return this
    }

    fun build() {
        context!!
        ILog.initialize(logAbstract)
        if (mIImageLoaderProxy != null) {
            ImageLoader.instance
                    .setImageLoaderProxy(mIImageLoaderProxy)
        }
        if (developConfig != null) {
            DevelopMode
                    .getInstance()
                    .InitializationConfig(developConfig)
        }
        MMKV.initialize(context)
        if (spProxy != null) {
            SPUtils.initialize(spProxy)
        } else {
            mMMKV = MMKV.mmkvWithID(context!!.packageName)
            SPUtils.initialize(object : SPProxy {
                override fun filter(key: String?): MMKV? {
                    return mMMKV
                }
            })
        }
    }

    companion object {
        private var manager: BasicOptions? = null

        @JvmStatic
        val instance: BasicOptions
            get() {
                if (manager == null) {
                    synchronized(BasicOptions::class.java) {
                        manager = BasicOptions()
                    }
                }
                return manager!!
            }
    }
}